Migrate EmojiPickerMenu to FlashList

TMisiukiewicz opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
Migrating EmojiPickerMenu to FlashList for both Web and Mobile
### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/30911
PROPOSAL: 


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Open emoji picker
2. Scroll the list
3. Press the emojis category on the top and verify it scrolls to the category
4. Change skin tone and verify it changed in the emoji list
5. Type "aaa" in search bar and verify no results message is displayed
6. Type "a" in search bar and verify filtered list scrolls 
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
n/a
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
9. Upload an image via copy paste
10. Verify a modal appears displaying a preview of that image
--->
1. Open emoji picker
2. Scroll the list
3. Press the emojis category on the top and verify it scrolls to the category
4. Change skin tone and verify it changed in the emoji list
5. Type "aaa" in search bar and verify no results message is displayed
6. Type "a" in search bar and verify filtered list scrolls 
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/13985840/2902d724-2742-49dd-bb9c-a0e8de62f3b3


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/13985840/49f00a4e-c35b-4f5b-b397-0d000a286635


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/13985840/da70eadf-31e6-4c5e-9536-54f3260bb499


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/13985840/1201562f-28ec-444c-90d7-58a72ade41d0


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/13985840/ec91e288-4aca-491b-8460-c66d6650870d


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/13985840/93677f6c-28bc-455f-8d30-5821674b6b47


</details>

barttom commented on src/components/EmojiPicker/EmojiPickerMenu/index.js:
```diff
@@ -421,15 +407,15 @@ function EmojiPickerMenu(props) {
             if (header) {
                 return (
-                    <View style={styles.emojiHeaderContainer}>
+                    <View style={[styles.emojiHeaderContainer, target === 'StickyHeader' ? styles.mh4 : {}]}>
```
wondering if can by just undefined instead of an empty object?

barttom commented on src/components/EmojiPicker/EmojiPickerMenu/index.native.js:
```diff
@@ -149,6 +147,24 @@ function EmojiPickerMenu({preferredLocale, onEmojiSelected, preferredSkinTone, t
+     * @param {Object} item
+     * @returns {String}
+     */
+    const getItemType = (item) => {
```
I've seen similar function in EmojiPickerMenu - could YOu share just one function?

fabioh8010 commented on src/components/EmojiPicker/EmojiPickerMenu/index.js:
```diff
@@ -384,7 +371,6 @@ function EmojiPickerMenu(props) {
         }
 
         const calculatedOffset = Math.floor(headerIndex / CONST.EMOJI_NUM_PER_ROW) * CONST.EMOJI_PICKER_HEADER_HEIGHT;
-        emojiListRef.current.flashScrollIndicators();
```
Why was this line removed?

TMisiukiewicz replied:
`flashScrollIndicators` method is not supported in FlashList

fabioh8010 commented on src/components/EmojiPicker/EmojiPickerMenu/index.native.js:
```diff
@@ -100,7 +99,6 @@ function EmojiPickerMenu({preferredLocale, onEmojiSelected, preferredSkinTone, t
 
     const scrollToHeader = (headerIndex) => {
         const calculatedOffset = Math.floor(headerIndex / CONST.EMOJI_NUM_PER_ROW) * CONST.EMOJI_PICKER_HEADER_HEIGHT;
-        emojiList.current.flashScrollIndicators();
```
same

cubuspl42 commented on src/components/EmojiPicker/EmojiPickerMenu/index.js:
```diff
@@ -470,61 +457,67 @@ function EmojiPickerMenu(props) {
-                renderItem={renderItem}
-                keyExtractor={keyExtractor}
-                numColumns={CONST.EMOJI_NUM_PER_ROW}
+        <View>
```
What's happening here? Why we nest a `View` inside a `View` with no style or attributes?

TMisiukiewicz replied:
good catch, thank you!

cubuspl42 commented:
Exception:

https://github.com/Expensify/App/assets/2590174/dacab996-5dda-4525-9fce-f2285c59d873

TMisiukiewicz commented:
@cubuspl42 thanks for testing, I addressed both of your comments

TMisiukiewicz commented:
hi @cubuspl42, did you have a chance to take another look on this PR?

cubuspl42 commented:
@TMisiukiewicz I will; I have a personal situation that affects my capacity. Everything will be back to normal after Friday. Is this urgent or conflict-prone?

TMisiukiewicz commented:
@cubuspl42 no worries, it can wait until you are back!

TMisiukiewicz commented:
hi @cubuspl42, will you find some time to check this PR this week? 🙏

cubuspl42 commented on src/components/EmojiPicker/EmojiPickerMenu/getItemType.js:
```diff
@@ -0,0 +1,20 @@
+/**
```
We don't allow new JavaScript files, this should be ported to TypeScript

TMisiukiewicz replied:
[This](https://expensify.slack.com/archives/C01GTK53T8Q/p1700075112873269) announcement doesn't mention `src/components`. Should I do that anyway? It would require typing the `item`

cubuspl42 commented:
Please also merge `main`

cubuspl42 commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/2590174/17961475-ee11-42e0-8e27-c32e15638fba

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/2590174/a9273e00-87a3-4da1-869d-982623712124

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/2590174/448cb4f2-0a46-439b-a1e1-851f48782015

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/2590174/149efeca-d51c-45a6-bc85-5511ad389b50

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/2590174/547eff39-d4b8-4bc5-82d7-056c500d1df8

</details>

jakub-trzebiatowski reviewed (changes requested):
Left one comment, otherwise things look good!

TMisiukiewicz commented:
`main` has been merged ✅

jakub-trzebiatowski reviewed (approved):
👍

roryabraham reviewed (changes requested):
Also please investigate what can be DRYed up between the web and mobile implementations, creating a Base component if necessary

TMisiukiewicz commented on src/components/EmojiPicker/EmojiPickerMenu/getItemType.js:
```diff
@@ -0,0 +1,20 @@
+        return;
+    }
+    if (item.header) {
+        return 'header';
```
According to the FlashList docs, this method is called very often so it should be as fast as possible. I'd stick to if statement here instead of looping through an array.

Created a constant for available types ✅

roryabraham commented:
Conflicts and lint errors

TMisiukiewicz commented:
@roryabraham done ✅

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/BaseEmojiPickerMenu.js:
```diff
@@ -0,0 +1,139 @@
+import getItemType from './getItemType';
+
+const emojiPropType = {
+    code: PropTypes.string.isRequired,
```
Please include JSDoc style comments for each of these props

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/BaseEmojiPickerMenu.js:
```diff
@@ -0,0 +1,139 @@
+    preferredSkinTone,
+    extraData,
+    alwaysBounceVertical,
+    translate,
```
Please use the `useLocalize` hook instead of `withLocalize` HOC

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/getItemType.js:
```diff
@@ -0,0 +1,22 @@
+    if (item.spacer) {
+        return CONST.EMOJI_PICKER_ITEM_TYPES.SPACER;
+    }
+    return CONST.EMOJI_PICKER_ITEM_TYPES.EMOJI;
```
If you're really trying to hyper-optimize this we should check this (the most common case) before checking for headers and spacers

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/BaseEmojiPickerMenu.js:
```diff
@@ -0,0 +1,139 @@
+import stylePropTypes from '@styles/stylePropTypes';
+import useThemeStyles from '@styles/useThemeStyles';
+import CONST from '@src/CONST';
+import getItemType from './getItemType';
```
I think we can move getItemType into this file now

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/index.js:
```diff
@@ -47,6 +46,31 @@ const defaultProps = {
+ * @param {Number} index
+ * @returns {String}
+ */
+const keyExtractor = (item, index) => `emoji_picker_${item.code}_${index}`;
```
keyExtractor looks like it can be moved to the base implementation file

TMisiukiewicz commented:
@roryabraham I applied all requested changes

roryabraham commented:
Sorry I didn't review this earlier as I've had some unplanned ooo. Can you please resolve conflicts here?

roryabraham reviewed (changes requested):
I also think there's a great deal more that can be done to DRY up the implementations of `EmojiPickerMenu/index.js` and `EmojiPickerMenu/index.native.js`. For example, in `EmojiPickerMenu/index.js`, we use an over-complicated combination of `useRef`, `useState`, and `useEffect` to accomplish what we do more simply with `useMemo` in `EmojiPickerMenu/index.js`.

```diff
diff --git a/assets/emojis/index.ts b/assets/emojis/index.ts
index 9f73fcf290..3c2849b41e 100644
--- a/assets/emojis/index.ts
+++ b/assets/emojis/index.ts
@@ -33,15 +33,21 @@ const localeEmojis = {
     es: esEmojis,
 } as const;
 
-const flagHeaderIndex = emojis.findIndex((emoji) => {
-    if (!('header' in emoji)) {
-        return;
-    }
-
-    return emoji.header && emoji.code === 'flags';
-});
-
-const emojisForOperatingSystem = getOperatingSystem() === CONST.OS.WINDOWS ? emojis.slice(0, flagHeaderIndex) : emojis;
-
+// On windows, flag emojis are not supported
+const emojisForOperatingSystem =
+    getOperatingSystem() === CONST.OS.WINDOWS
+        ? emojis.slice(
+              0,
+              emojis.findIndex((emoji) => {
+                  if (!('header' in emoji)) {
+                      return;
+                  }
+
+                  return emoji.header && emoji.code === 'flags';
+              }),
+          )
+        : emojis;
+
+export default emojisForOperatingSystem;
 export {emojiNameTable, emojiCodeTableWithSkinTones, localeEmojis, emojisForOperatingSystem};
-export {skinTones, categoryFrequentlyUsed, default} from './common';
+export {skinTones, categoryFrequentlyUsed} from './common';
diff --git a/src/components/EmojiPicker/EmojiPickerMenu/index.js b/src/components/EmojiPicker/EmojiPickerMenu/index.js
index b3d4cdb00f..5affc452b9 100755
--- a/src/components/EmojiPicker/EmojiPickerMenu/index.js
+++ b/src/components/EmojiPicker/EmojiPickerMenu/index.js
@@ -1,18 +1,17 @@
 import lodashGet from 'lodash/get';
 import PropTypes from 'prop-types';
-import React, {useCallback, useEffect, useRef, useState} from 'react';
+import React, {useCallback, useEffect, useMemo, useRef, useState} from 'react';
 import {View} from 'react-native';
 import {withOnyx} from 'react-native-onyx';
 import _ from 'underscore';
-import {emojisForOperatingSystem} from '@assets/emojis';
+import emojis from '@assets/emojis';
 import EmojiPickerMenuItem from '@components/EmojiPicker/EmojiPickerMenuItem';
 import Text from '@components/Text';
 import TextInput from '@components/TextInput';
-import withLocalize, {withLocalizePropTypes} from '@components/withLocalize';
+import useLocalize from '@hooks/useLocalize';
 import useWindowDimensions from '@hooks/useWindowDimensions';
 import * as Browser from '@libs/Browser';
 import canFocusInputOnScreenFocus from '@libs/canFocusInputOnScreenFocus';
-import compose from '@libs/compose';
 import * as EmojiUtils from '@libs/EmojiUtils';
 import isEnterWhileComposition from '@libs/KeyboardShortcut/isEnterWhileComposition';
 import * as ReportUtils from '@libs/ReportUtils';
@@ -35,8 +34,6 @@ const propTypes = {
     /** Stores user's frequently used emojis */
     // eslint-disable-next-line react/forbid-prop-types
     frequentlyUsedEmojis: PropTypes.arrayOf(PropTypes.object),
-
-    ...withLocalizePropTypes,
 };
 
 const defaultProps = {
@@ -47,24 +44,8 @@ const defaultProps = {
 
 const throttleTime = Browser.isMobile() ? 200 : 50;
 
-/**
- * Calculate the filtered + header emojis and header row indices
- * @returns {Object}
- */
-function getEmojisAndHeaderRowIndices() {
-    const filteredEmojis = EmojiUtils.mergeEmojisWithFrequentlyUsedEmojis(emojisForOperatingSystem);
-    // Get the header emojis along with the code, index and icon.
-    // index is the actual header index starting at the first emoji and counting each one
-    const headerEmojis = EmojiUtils.getHeaderEmojis(filteredEmojis);
-
-    // FlashList renders items defined as null, so we have to store each header's row index to set their style properly
-    const headerRowIndices = _.map(headerEmojis, (headerEmoji) => headerEmoji.index);
-
-    return {filteredEmojis, headerEmojis, headerRowIndices};
-}
-
-function EmojiPickerMenu(props) {
-    const {forwardedRef, frequentlyUsedEmojis, preferredSkinTone, onEmojiSelected, preferredLocale, translate} = props;
+function EmojiPickerMenu({forwardedRef, frequentlyUsedEmojis, preferredSkinTone, onEmojiSelected}) {
+    const {preferredLocale, translate} = useLocalize();
 
     const styles = useThemeStyles();
 
@@ -80,14 +61,16 @@ function EmojiPickerMenu(props) {
     // prevent auto focus when open picker for mobile device
     const shouldFocusInputOnScreenFocus = canFocusInputOnScreenFocus();
 
-    const firstNonHeaderIndex = useRef(0);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    const allEmojis = useMemo(() => EmojiUtils.mergeEmojisWithFrequentlyUsedEmojis(emojis), [frequentlyUsedEmojis]);
+    const headerEmojis = useMemo(() => EmojiUtils.getHeaderEmojis(allEmojis), [allEmojis]);
+    const headerRowIndices = useMemo(() => _.map(headerEmojis, (headerEmoji) => headerEmoji.index), [headerEmojis]);
 
-    const [emojis, setEmojis] = useState([]);
-    const [initialHeaderRowIndices, setInitialHeaderRowIndices] = useState([]);
-    const [headerEmojis, setHeaderEmojis] = useState([]);
+    const [filteredEmojis, setFilteredEmojis] = useState(allEmojis);
+    const [headerIndices, setHeaderIndices] = useState(headerRowIndices);
+
+    const firstNonHeaderIndex = useMemo(() => _.findIndex(filteredEmojis, (item) => !item.spacer && !item.header), [filteredEmojis]);
 
-    const [filteredEmojis, setFilteredEmojis] = useState([]);
-    const [headerIndices, setHeaderIndices] = useState([]);
     const [highlightedIndex, setHighlightedIndex] = useState(-1);
     const [arePointerEventsDisabled, setArePointerEventsDisabled] = useState(false);
     const [selection, setSelection] = useState({start: 0, end: 0});
@@ -96,15 +79,12 @@ function EmojiPickerMenu(props) {
     const [highlightFirstEmoji, setHighlightFirstEmoji] = useState(false);
 
     useEffect(() => {
-        const emojisAndHeaderRowIndices = getEmojisAndHeaderRowIndices();
-        if (emojis.length === 0 || initialHeaderRowIndices.length === 0) {
-            setEmojis(emojisAndHeaderRowIndices.filteredEmojis);
-            setInitialHeaderRowIndices(emojisAndHeaderRowIndices.headerRowIndices);
-        }
-        setHeaderEmojis(emojisAndHeaderRowIndices.headerEmojis);
-        setFilteredEmojis(emojisAndHeaderRowIndices.filteredEmojis);
-        setHeaderIndices(emojisAndHeaderRowIndices.headerRowIndices);
-    }, [emojis.length, frequentlyUsedEmojis, initialHeaderRowIndices.length]);
+        setFilteredEmojis(allEmojis);
+    }, [allEmojis]);
+
+    useEffect(() => {
+        setHeaderIndices(headerRowIndices);
+    }, [headerRowIndices]);
 
     /**
      * On text input selection change
@@ -115,14 +95,6 @@ function EmojiPickerMenu(props) {
         setSelection(event.nativeEvent.selection);
     }, []);
 
-    /**
-     * Find and store index of the first emoji item
-     * @param {Array} filteredEmojisArr
-     */
-    function updateFirstNonHeaderIndex(filteredEmojisArr) {
-        firstNonHeaderIndex.current = _.findIndex(filteredEmojisArr, (item) => !item.spacer && !item.header);
-    }
-
     const mouseMoveHandler = useCallback(() => {
         if (!arePointerEventsDisabled) {
             return;
@@ -147,10 +119,9 @@ function EmojiPickerMenu(props) {
         }
         if (normalizedSearchTerm === '') {
             // There are no headers when searching, so we need to re-make them sticky when there is no search term
-            setFilteredEmojis(emojis);
-            setHeaderIndices(initialHeaderRowIndices);
+            setFilteredEmojis(allEmojis);
+            setHeaderIndices(headerRowIndices);
             setHighlightedIndex(-1);
-            updateFirstNonHeaderIndex(emojis);
             setHighlightFirstEmoji(false);
             return;
         }
@@ -160,7 +131,6 @@ function EmojiPickerMenu(props) {
         setFilteredEmojis(newFilteredEmojiList);
         setHeaderIndices([]);
         setHighlightedIndex(0);
-        updateFirstNonHeaderIndex(newFilteredEmojiList);
         setHighlightFirstEmoji(true);
     }, throttleTime);
 
@@ -200,7 +170,7 @@ function EmojiPickerMenu(props) {
             // If nothing is highlighted and an arrow key is pressed
             // select the first emoji, apply keyboard movement styles, and disable pointer events
             if (highlightedIndex === -1) {
-                setHighlightedIndex(firstNonHeaderIndex.current);
+                setHighlightedIndex(firstNonHeaderIndex);
                 setArePointerEventsDisabled(true);
                 setIsUsingKeyboardMovement(true);
                 return;
@@ -230,7 +200,7 @@ function EmojiPickerMenu(props) {
                 case 'ArrowLeft':
                     move(
                         -1,
-                        () => highlightedIndex - 1 < firstNonHeaderIndex.current,
+                        () => highlightedIndex - 1 < firstNonHeaderIndex,
                         () => {
                             // Reaching start of the list, arrow left set the focus to searchInput.
                             focusInputWithTextSelect();
@@ -244,7 +214,7 @@ function EmojiPickerMenu(props) {
                 case 'ArrowUp':
                     move(
                         -CONST.EMOJI_NUM_PER_ROW,
-                        () => highlightedIndex - CONST.EMOJI_NUM_PER_ROW < firstNonHeaderIndex.current,
+                        () => highlightedIndex - CONST.EMOJI_NUM_PER_ROW < firstNonHeaderIndex,
                         () => {
                             // Reaching start of the list, arrow up set the focus to searchInput.
                             focusInputWithTextSelect();
@@ -349,11 +319,6 @@ function EmojiPickerMenu(props) {
         };
     }, [forwardedRef, shouldFocusInputOnScreenFocus, cleanupEventHandlers, setupEventHandlers]);
 
-    useEffect(() => {
-        // Find and store index of the first emoji item on mount
-        updateFirstNonHeaderIndex(emojis);
-    }, [emojis]);
-
     const scrollToHeader = useCallback((headerIndex) => {
         if (!emojiListRef.current) {
             return;
@@ -417,7 +382,7 @@ function EmojiPickerMenu(props) {
         [preferredSkinTone, highlightedIndex, isUsingKeyboardMovement, highlightFirstEmoji, styles, translate, onEmojiSelected],
     );
 
-    const isFiltered = emojis.length !== filteredEmojis.length;
+    const isFiltered = allEmojis.length !== filteredEmojis.length;
     const listStyle = StyleUtils.getEmojiPickerListHeight(isFiltered, windowHeight);
     const height = !listStyle.maxHeight || listStyle.height < listStyle.maxHeight ? listStyle.height : listStyle.maxHeight;
     const overflowLimit = Math.floor(height / CONST.EMOJI_PICKER_ITEM_HEIGHT) * 8;
@@ -489,14 +454,11 @@ const EmojiPickerMenuWithRef = React.forwardRef((props, ref) => (
 
 EmojiPickerMenuWithRef.displayName = 'EmojiPickerMenuWithRef';
 
-export default compose(
-    withLocalize,
-    withOnyx({
-        preferredSkinTone: {
-            key: ONYXKEYS.PREFERRED_EMOJI_SKIN_TONE,
-        },
-        frequentlyUsedEmojis: {
-            key: ONYXKEYS.FREQUENTLY_USED_EMOJIS,
-        },
-    }),
-)(EmojiPickerMenuWithRef);
+export default withOnyx({
+    preferredSkinTone: {
+        key: ONYXKEYS.PREFERRED_EMOJI_SKIN_TONE,
+    },
+    frequentlyUsedEmojis: {
+        key: ONYXKEYS.FREQUENTLY_USED_EMOJIS,
+    },
+})(EmojiPickerMenuWithRef);
diff --git a/src/components/EmojiPicker/EmojiPickerMenu/index.native.js b/src/components/EmojiPicker/EmojiPickerMenu/index.native.js
index e990c71e99..778382c18b 100644
--- a/src/components/EmojiPicker/EmojiPickerMenu/index.native.js
+++ b/src/components/EmojiPicker/EmojiPickerMenu/index.native.js
@@ -8,10 +8,9 @@ import emojis from '@assets/emojis';
 import EmojiPickerMenuItem from '@components/EmojiPicker/EmojiPickerMenuItem';
 import Text from '@components/Text';
 import TextInput from '@components/TextInput';
-import withLocalize, {withLocalizePropTypes} from '@components/withLocalize';
+import useLocalize from '@hooks/useLocalize';
 import useSingleExecution from '@hooks/useSingleExecution';
 import useWindowDimensions from '@hooks/useWindowDimensions';
-import compose from '@libs/compose';
 import * as EmojiUtils from '@libs/EmojiUtils';
 import * as StyleUtils from '@styles/StyleUtils';
 import useThemeStyles from '@styles/useThemeStyles';
@@ -30,9 +29,6 @@ const propTypes = {
     /** Stores user's frequently used emojis */
     // eslint-disable-next-line react/forbid-prop-types
     frequentlyUsedEmojis: PropTypes.arrayOf(PropTypes.object),
-
-    /** Props related to translation */
-    ...withLocalizePropTypes,
 };
 
 const defaultProps = {
@@ -40,17 +36,20 @@ const defaultProps = {
     frequentlyUsedEmojis: [],
 };
 
-function EmojiPickerMenu({preferredLocale, onEmojiSelected, preferredSkinTone, translate, frequentlyUsedEmojis}) {
+function EmojiPickerMenu({frequentlyUsedEmojis, preferredSkinTone, onEmojiSelected}) {
+    const {preferredLocale, translate} = useLocalize();
     const styles = useThemeStyles();
     const emojiList = useAnimatedRef();
+    const {windowWidth} = useWindowDimensions();
+    const {singleExecution} = useSingleExecution();
+
     // eslint-disable-next-line react-hooks/exhaustive-deps
     const allEmojis = useMemo(() => EmojiUtils.mergeEmojisWithFrequentlyUsedEmojis(emojis), [frequentlyUsedEmojis]);
     const headerEmojis = useMemo(() => EmojiUtils.getHeaderEmojis(allEmojis), [allEmojis]);
     const headerRowIndices = useMemo(() => _.map(headerEmojis, (headerEmoji) => headerEmoji.index), [headerEmojis]);
+
     const [filteredEmojis, setFilteredEmojis] = useState(allEmojis);
     const [headerIndices, setHeaderIndices] = useState(headerRowIndices);
-    const {windowWidth} = useWindowDimensions();
-    const {singleExecution} = useSingleExecution();
 
     useEffect(() => {
         setFilteredEmojis(allEmojis);
@@ -178,14 +177,11 @@ const EmojiPickerMenuWithRef = React.forwardRef((props, ref) => (
 
 EmojiPickerMenuWithRef.displayName = 'EmojiPickerMenuWithRef';
 
-export default compose(
-    withLocalize,
-    withOnyx({
-        preferredSkinTone: {
-            key: ONYXKEYS.PREFERRED_EMOJI_SKIN_TONE,
-        },
-        frequentlyUsedEmojis: {
-            key: ONYXKEYS.FREQUENTLY_USED_EMOJIS,
-        },
-    }),
-)(EmojiPickerMenuWithRef);
+export default withOnyx({
+    preferredSkinTone: {
+        key: ONYXKEYS.PREFERRED_EMOJI_SKIN_TONE,
+    },
+    frequentlyUsedEmojis: {
+        key: ONYXKEYS.FREQUENTLY_USED_EMOJIS,
+    },
+})(EmojiPickerMenuWithRef);
```

roryabraham reviewed (commented):
Can you also investigate our implementation of highlighting / arrow key navigation in `EmojiPickerMenu/index.js` and see if we can simplify/standardize it a bit by leveraging the `useArrowKeyFocusManager` hook?

roryabraham commented:
Also, I strongly recommend using a file comparison tool (I use WebStorm), to look at the diff between `EmojiPickerMenu/index.js` and `EmojiPickerMenu/index.native.js` side-by-side, and iteratively review the diff between the files to see what differences are necessary and what can be potentially DRYed up:

<img width="1509" alt="image" src="https://github.com/Expensify/App/assets/47436092/410fe724-0a32-452a-a829-5fd80e9d2530">

cubuspl42 commented:
@roryabraham Good catch with the repeated code. I'll start using the IDE diffing trick 👍

TMisiukiewicz commented:
@roryabraham thanks for your valuable comments 🙌 I already applied most of them. 

I did an investigation about `useArrowKeyFocusManager`, however it currently supports only ArrowUp and ArrowDown, and we need the ArrowLeft and ArrowRight as well. I'd say for now we should stick to the current implementation, because mixing both implementations would be confusing. We could think of adding  ArrowLeft and ArrowRight to `useArrowKeyFocusManager`, but I'd treat it as something to tackle separately from this task.

I'll do another round of review what could be possibly DRYed up tomorrow - will let you know once it's done

TMisiukiewicz commented:
@roryabraham it's ready for review

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/BaseEmojiPickerMenu.js:
```diff
@@ -0,0 +1,146 @@
+                    keyExtractor={keyExtractor}
+                    numColumns={CONST.EMOJI_NUM_PER_ROW}
+                    stickyHeaderIndices={stickyHeaderIndices}
+                    ListEmptyComponent={<Text style={[styles.textLabel, styles.colorMuted]}>{translate('common.noResultsFound')}</Text>}
```
@adhorodyski can you please review [this discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1681459454343289) for context and then let us know if you still think we should make a change here?

TMisiukiewicz replied:
@adhorodyski I applied the following change after internal discussion, could you please share your thoughts on this here as well?

adhorodyski replied:
Sure thing!
Since Flashlist extends from a PureComponent, by passing in `EmptyComponent={<Text />}` instead of `EmptyComponent={Empty}` we create a new React element with each render, thus breaking the memoisation (a new object reference - in this case `<Text />` behaves exactly like passing in a lambda expression by resulting in a React.createElement call, see https://react.dev/reference/react/createElement).

This is bad performance wise - we not only allocate more memory (N nodes VS 1 node, more GC needed), but also force a rerender on our list - all the comparisons that were made (CPU cycles spent on comparing the props) were a waste. 

The optimisation efforts (like wrapping other props with useCallback etc) lead to nothing if we break the memoisation by passing in a reference to a new object for a single property instead of reusing a component instance.

@roryabraham please let me know if this makes things more clear or I should rephrase something! Happy to chat about it 🙂

adhorodyski replied:
TLDR: for pure/memoised components `<Text />` behaves like `() =>` called internally - a new object (element) is getting returned.

cubuspl42 replied:
I was confused by this thread, as typically you can't _just like that_ exchange `<Foo>` and `Foo` in React, but then I realized that the type of `ListEmptyComponent` is a union; it accepts both.

```ts
ListEmptyComponent?: React.ComponentType<any> | React.ReactElement<any, string | React.JSXElementConstructor<any>>;
```

cubuspl42 replied:
> a new object (element) is getting returned.

It seems that in this case, we wouldn't trigger extraneous renders, but I guess we would put more stress on the diffing system

roryabraham replied:
ok, I think that makes sense. Let's change it back to pass a reference to the component instead of an element directly then.

roryabraham commented:
Thanks for all the work here and sorry for the delay in re-reviewing @TMisiukiewicz, jumping back into this PR review now

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/index.native.js:
```diff
@@ -1,76 +1,41 @@
-    }, [headerRowIndices]);
-
-    const getItemLayout = (data, index) => ({length: CONST.EMOJI_PICKER_ITEM_HEIGHT, offset: CONST.EMOJI_PICKER_ITEM_HEIGHT * index, index});
+    const emojiList = useAnimatedRef();
```
I think we _might_ be able to use `useAnimatedRef` and `scrollTo` on web as well, so that could be another difference between these files that could be eliminated.

When we follow-up to add `ArrowRight` and `ArrowLeft` handling to `useArrowKeyFocusManager`, then the ref and imperative scrolling logic can be moved into `BaseEmojiPickerMenu`.

roryabraham reviewed (changes requested):
Getting much better, thanks for taking the time to make this code so much better.

TMisiukiewicz committed:
initial migration to flashlist

TMisiukiewicz committed:
fix displaying headers

TMisiukiewicz committed:
display emojis correctly with flashlist

TMisiukiewicz committed:
sticky headers in emoji picker

TMisiukiewicz committed:
correctly display changing skin tone

TMisiukiewicz committed:
scroll to offset

TMisiukiewicz committed:
scroll to offset

TMisiukiewicz committed:
display headers on native correctly

TMisiukiewicz committed:
display scrollbar

TMisiukiewicz committed:
update getItemType

TMisiukiewicz committed:
create separate file for getItemType

TMisiukiewicz committed:
slice flags for operating system in asset

TMisiukiewicz committed:
code review updates

TMisiukiewicz committed:
update getItemType and list style

TMisiukiewicz committed:
extract updatePreferredSkinTone function

TMisiukiewicz committed:
create BaseEmojiPickerMenu

TMisiukiewicz committed:
code review updates

TMisiukiewicz committed:
update BaseEmojiPickerMenu

TMisiukiewicz committed:
code review updates

TMisiukiewicz committed:
Revert "implement useArrowKeyFocusManager"

This reverts commit 646bbb1602a1265f22ffdb8fe23887f2d54fd49b.

TMisiukiewicz committed:
implement useArrowKeyFocusManager

TMisiukiewicz committed:
refactor EmojiPickerMenu for web

TMisiukiewicz committed:
simplify implementation across platforms for emoji picker menu

TMisiukiewicz committed:
get empty list component out of render loop

TMisiukiewicz commented on src/components/EmojiPicker/EmojiPickerMenu/index.js:
```diff
@@ -468,8 +357,7 @@ function EmojiPickerMenu(props) {
-    const isFiltered = emojis.current.length !== filteredEmojis.length;
-    const listStyle = StyleUtils.getEmojiPickerListHeight(isFiltered, windowHeight);
+    const listStyle = StyleUtils.getEmojiPickerListHeight(isListFiltered, windowHeight);
     const height = !listStyle.maxHeight || listStyle.height < listStyle.maxHeight ? listStyle.height : listStyle.maxHeight;
```
since we removed `overflowY`, height is not necessary anymore. Removing it entirely

TMisiukiewicz committed:
code review updates

TMisiukiewicz commented:
@roryabraham thanks, applied all the necessary changes! Additionally, when testing I noticed that there is a barely-visible flicker of the list when opening the Emoji picker. It's probably caused by missing `estimatedListSize` prop in the Flashlist. I'll investigate and fix it on Monday

roryabraham commented:
@TMisiukiewicz Please try not to force-push in general while working on code for Expensify as it makes things harder to review. Feel free to just merge main into your branch rather than rebasing

roryabraham commented on src/styles/utils/index.ts:
```diff
@@ -847,15 +846,14 @@ function displayIfTrue(condition: boolean): ViewStyle {
  * Gets the correct height for emoji picker list based on screen dimensions
  */
-function getEmojiPickerListHeight(hasAdditionalSpace: boolean, windowHeight: number): ViewStyle {
+function getEmojiPickerListHeight(isRenderingShortcutRow: boolean, windowHeight: number): ViewStyle {
```
This function appears unused now:

<img width="1401" alt="image" src="https://github.com/Expensify/App/assets/47436092/bb6018cd-4caa-4072-b159-d38345957e6b">

let's delete it from this file accordingly

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/index.js:
```diff
@@ -485,7 +396,7 @@ function EmojiPickerMenu(props) {
                     label={translate('common.search')}
                     accessibilityLabel={translate('common.search')}
-                    role={CONST.ROLE.PRESENTATION}
+                    role={CONST.ACCESSIBILITY_ROLE.TEXT}
```
This looks like a bad merge – should be `CONST.ACCESSIBILITY_ROLE` is deprecated and `CONST.ROLE` should be used instead

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/index.js:
```diff
@@ -501,41 +412,27 @@ function EmojiPickerMenu(props) {
+                listWrapperStyle={[
                     listStyle,
-                    // This prevents elastic scrolling when scroll reaches the start or end
-                    {overscrollBehaviorY: 'contain'},
```
Can we be sure to retest the [original bug](https://github.com/Expensify/App/issues/22939#issuecomment-1655555722) in iOS Safari? Apparently that's where it occurs: https://expensify.slack.com/archives/C01GTK53T8Q/p1702629684965709?thread_ts=1702597102.525129&cid=C01GTK53T8Q

TMisiukiewicz replied:
I couldn't reproduce it - seems like it's safe to remove it

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/index.native.js:
```diff
@@ -126,31 +78,32 @@ function EmojiPickerMenu({preferredLocale, onEmojiSelected, preferredSkinTone, t
+    );
 
     return (
         <View style={styles.emojiPickerContainer}>
```
Can you check what happens if we also pass `StyleUtils.getEmojiPickerStyle(isSmallScreenWidth)` to this view style on native? It seems like it should be safe to pass on both web and native since its return value is determined by `isSmallScreenWidth`

TMisiukiewicz replied:
Checked it and looks like it's safe to add it here ✅

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/index.native.js:
```diff
@@ -1,79 +1,53 @@
-        if (emojiList.current) {
-            emojiList.current.scrollToOffset({offset: 0, animated: false});
+        if (emojiListRef.current) {
+            emojiListRef.current.scrollToOffset({offset: 0, animated: false});
```
I think we can use `scrollTo` here, same as we do in `index.js`

TMisiukiewicz replied:
I'm noticing a small flickering on mobile here when I am using `scrollTo` - it looks like when using it, the scroll to top is faster than filtering the list as it's not animating - list is scrolled to top first and after a few ms data is overwritten with new data. Not sure why, but it works correct when using the approach above. Attaching a video of the issue when using `scrollTo` here.

https://github.com/Expensify/App/assets/13985840/ee9bfe88-0ba1-40f4-99ad-3b49a5363185

roryabraham commented on src/components/EmojiPicker/EmojiPickerMenu/emojiPickerMenuPropTypes.js:
```diff
@@ -0,0 +1,21 @@
+    onEmojiSelected: PropTypes.func.isRequired,
+
+    /** Stores user's preferred skin tone */
+    preferredSkinTone: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
```
Looks like, thanks to the hooks you added, the only prop still needed for `EmojiPickerMenu` is `onEmojiSelected`. So let's remove these other two along with their defaults

TMisiukiewicz committed:
Merge branch 'main' into feat/emojipicker-flashlist-migration

TMisiukiewicz committed:
code review updates

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into feat/emojipicker-flashlist-migration

TMisiukiewicz committed:
add estimatedListSize to emoji picker menu

TMisiukiewicz committed:
update getEmojiPickerListHeight

TMisiukiewicz commented:
@roryabraham thanks for your eagle-eye reviews! Really appreciate all the stuff you keep findind and your knowledge of the project. I addressed your comments and fixed the flicker on the initial list render I mentioned on Friday

roryabraham reviewed (approved):
Thanks for all your work on this. It's much better than it was. I do want us to follow-up with a PR to build support for `ArrowLeft` and `ArrowRight` in `useArrowKeyFocusManager` as that will further reduce platform differences between these files, I believe to the point where everything can be recombined into a single component.

roryabraham commented:
@cubuspl42 there have been a lot of changes here since your last review. Can you please review and test again before I merge this? I see the reviewer checklist is failing, maybe try again with a fresh checklist?

TMisiukiewicz commented:
thanks for your review @roryabraham! Can we create an issue for supporting `ArrowLeft` and `ArrowRight` in `useArrowKeyFocusManager`? Christmas break is very close and just want to make sure once we are back from holidays, me or other Callstack folk could pick this up

cubuspl42 commented:
@TMisiukiewicz Would you merge `main`? We're >600 commits behind

TMisiukiewicz committed:
Merge branch 'main' into feat/emojipicker-flashlist-migration

TMisiukiewicz commented:
@cubuspl42 whoops that's a pretty big number, updated!

cubuspl42 commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/2590174/17961475-ee11-42e0-8e27-c32e15638fba

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/2590174/a9273e00-87a3-4da1-869d-982623712124

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/2590174/448cb4f2-0a46-439b-a1e1-851f48782015

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/2590174/149efeca-d51c-45a6-bc85-5511ad389b50

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/2590174/547eff39-d4b8-4bc5-82d7-056c500d1df8

</details>

jakub-trzebiatowski reviewed (approved):
I re-tested, and things still work cross-platform!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7325124647) to staging by https://github.com/roryabraham in version: 1.4.17-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|cancelled 🔪
🕸 web 🕸|success ✅

mountiny commented on assets/emojis/index.ts:
```diff
@@ -31,5 +33,21 @@ const localeEmojis = {
 
-export {emojiNameTable, emojiCodeTableWithSkinTones, localeEmojis};
-export {skinTones, categoryFrequentlyUsed, default} from './common';
+// On windows, flag emojis are not supported
+const emojisForOperatingSystem =
+    getOperatingSystem() === CONST.OS.WINDOWS
+        ? emojis.slice(
+              0,
+              emojis.findIndex((emoji) => {
+                  if (!('header' in emoji)) {
+                      return;
+                  }
+
+                  return emoji.header && emoji.code === 'flags';
+              }),
+          )
+        : emojis;
```
@TMisiukiewicz This was implemented during the time you have worked on this PR but we have added the support for flag emojis on Windows in this issues and pr https://github.com/Expensify/App/issues/31717

There was this issue raised here https://github.com/Expensify/App/issues/33593, if you could comment on it so I can assign you and you could make a follow up to remove this Windows specific code. thanks!

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7339899700) to production by https://github.com/mountiny in version: 1.4.17-8 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

rushatgabhane commented on src/components/EmojiPicker/EmojiPickerMenu/index.native.js:
```diff
@@ -126,34 +78,35 @@ function EmojiPickerMenu({preferredLocale, onEmojiSelected, preferredSkinTone, t
+
+            if (item.header) {
+                return (
+                    <View style={[styles.emojiHeaderContainer, target === 'StickyHeader' ? styles.mh4 : {width: windowWidth}]}>
```
Coming from https://github.com/Expensify/App/issues/33709#issuecomment-1871727444

We had a bug here scroll not working over header.
RCA: emoji header is wrapped in an absolute container and overlaps with scroll bar
